File: src/market-data/kline-sqlite-durability.js
import { randomUUID } from 'node:crypto';
import {
  closeSync,
  fsyncSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from 'node:fs';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';

const DEFAULT_BUSY_TIMEOUT_MS = 30_000;
const DEFAULT_LOCK_TIMEOUT_MS = 30_000;
const DEFAULT_LOCK_POLL_MS = 25;
const INVALID_LOCK_GRACE_MS = 5_000;

function sleepSync(ms) {
  const buffer = new SharedArrayBuffer(4);
  Atomics.wait(new Int32Array(buffer), 0, 0, ms);
}

function processIsAlive(pid) {
  if (!Number.isInteger(pid) || pid <= 0) return false;
  try {
    process.kill(pid, 0);
    return true;
  } catch (error) {
    return error?.code === 'EPERM';
  }
}

function readLockOwner(lockPath) {
  try {
    return JSON.parse(readFileSync(lockPath, 'utf8'));
  } catch {
    return null;
  }
}

function staleLockReason(lockPath, owner) {
  if (owner?.pid && !processIsAlive(Number(owner.pid))) {
    return 'owner_process_not_alive';
  }
  if (owner) return null;
  try {
    return Date.now() - statSync(lockPath).mtimeMs >= INVALID_LOCK_GRACE_MS
      ? 'invalid_owner_record'
      : null;
  } catch {
    return null;
  }
}

function sameLockOwner(expected, current) {
  if (expected?.token || current?.token) return expected?.token === current?.token;
  if (expected?.pid || current?.pid) {
    return Number(expected?.pid) === Number(current?.pid)
      && expected?.owner === current?.owner;
  }
  return expected === null && current === null;
}

function tryAcquireCleanupGate(lockPath) {
  const cleanupPath = `${lockPath}.cleanup`;
  const token = randomUUID();
  let fd;
  try {
    fd = openSync(cleanupPath, 'wx', 0o600);
    writeFileSync(fd, JSON.stringify({
      schema_version: 'kline_sqlite_cleanup_gate.v1',
      token,
      pid: process.pid,
      acquired_at: new Date().toISOString(),
    }));
    fsyncSync(fd);
    return { cleanupPath, fd, token };
  } catch (error) {
    if (fd !== undefined) {
      try { closeSync(fd); } catch {}
      try { unlinkSync(cleanupPath); } catch {}
    }
    if (error?.code === 'EEXIST') return null;
    throw error;
  }
}

function releaseCleanupGate(gate) {
  if (!gate) return;
  try {
    const owner = readLockOwner(gate.cleanupPath);
    if (owner?.token === gate.token) {
      try {
        unlinkSync(gate.cleanupPath);
      } catch (error) {
        if (error?.code !== 'ENOENT') throw error;
      }
    }
  } finally {
    try { closeSync(gate.fd); } catch {}
  }
}

function removeStaleLock(lockPath, observedOwner) {
  const gate = tryAcquireCleanupGate(lockPath);
  if (!gate) return false;
  try {
    const currentOwner = readLockOwner(lockPath);
    const reason = staleLockReason(lockPath, currentOwner);
    if (!reason || !sameLockOwner(observedOwner, currentOwner)) return false;

    const stalePath = `${lockPath}.stale-${Date.now()}-${process.pid}-${randomUUID()}`;
    try {
      renameSync(lockPath, stalePath);
    } catch (error) {
      if (error?.code === 'ENOENT') return false;
      throw error;
    }
    try {
      unlinkSync(stalePath);
    } catch (error) {
      if (error?.code !== 'ENOENT') throw error;
    }
    console.warn(`[kline-sqlite] removed stale writer lock (${reason}): ${lockPath}`);
    return true;
  } finally {
    releaseCleanupGate(gate);
  }
}

export function defaultKlineWriterLockPath(dbPath) {
  void dbPath;
  return join(tmpdir(), 'kline_sqlite_single_writer.lock');
}

export function configureKlineSqliteSync(db, options = {}) {
  const requestedJournalMode = String(
    options.journalMode || process.env.KLINE_SQLITE_JOURNAL_MODE || 'DELETE',
  ).trim().toUpperCase();
  if (requestedJournalMode !== 'DELETE') {
    throw new Error(`Kline SQLite requires journal_mode=DELETE; received ${requestedJournalMode}`);
  }

  const busyTimeoutMs = Math.max(
    1,
    Number(options.busyTimeoutMs || process.env.KLINE_SQLITE_BUSY_TIMEOUT_MS || DEFAULT_BUSY_TIMEOUT_MS),
  );
  db.pragma(`busy_timeout = ${Math.floor(busyTimeoutMs)}`);
  db.pragma('mmap_size = 0');
  const journalMode = String(db.pragma('journal_mode = DELETE', { simple: true })).toLowerCase();
  db.pragma('synchronous = FULL');
  const synchronous = Number(db.pragma('synchronous', { simple: true }));
  const mmapSize = Number(db.pragma('mmap_size', { simple: true }));

  if (journalMode !== 'delete' || synchronous !== 2 || mmapSize !== 0) {
    throw new Error(
      `Unsafe Kline SQLite settings: journal_mode=${journalMode} synchronous=${synchronous} mmap_size=${mmapSize}`,
    );
  }
  return {
    journal_mode: journalMode,
    synchronous: 'FULL',
    mmap_size: mmapSize,
    busy_timeout_ms: Math.floor(busyTimeoutMs),
  };
}

export function acquireKlineWriterLockSync(options = {}) {
  const lockPath = options.lockPath
    || process.env.KLINE_SQLITE_WRITER_LOCK_FILE
    || defaultKlineWriterLockPath(options.dbPath || 'default-kline-db');
  const timeoutMs = Math.max(1, Number(options.timeoutMs || DEFAULT_LOCK_TIMEOUT_MS));
  const pollMs = Math.max(1, Number(options.pollMs || DEFAULT_LOCK_POLL_MS));
  const token = randomUUID();
  const deadline = Date.now() + timeoutMs;
  mkdirSync(dirname(lockPath), { recursive: true });

  while (true) {
    let fd;
    try {
      fd = openSync(lockPath, 'wx', 0o600);
      writeFileSync(fd, JSON.stringify({
        schema_version: 'kline_sqlite_writer_lock.v1',
        token,
        pid: process.pid,
        owner: options.owner || 'unknown',
        acquired_at: new Date().toISOString(),
      }));
      fsyncSync(fd);
      return { fd, lockPath, token };
    } catch (error) {
      if (fd !== undefined) {
        try { closeSync(fd); } catch {}
      }
      if (error?.code !== 'EEXIST') throw error;

      const owner = readLockOwner(lockPath);
      const staleReason = staleLockReason(lockPath, owner);
      if (staleReason && removeStaleLock(lockPath, owner)) {
        continue;
      }
      if (Date.now() >= deadline) {
        const activeOwner = owner
          ? `pid=${owner.pid ?? 'unknown'} owner=${owner.owner ?? 'unknown'}`
          : 'owner=initializing';
        throw new Error(`Timed out waiting for Kline SQLite writer lock ${lockPath} (${activeOwner})`);
      }
      sleepSync(Math.min(pollMs, Math.max(1, deadline - Date.now())));
    }
  }
}

export function releaseKlineWriterLockSync(lock) {
  if (!lock) return;
  try {
    const owner = readLockOwner(lock.lockPath);
    if (owner?.token === lock.token) {
      try {
        unlinkSync(lock.lockPath);
      } catch (error) {
        if (error?.code !== 'ENOENT') throw error;
      }
    }
  } finally {
    try { closeSync(lock.fd); } catch {}
  }
}

export function withKlineWriterLockSync(owner, callback, options = {}) {
  const lock = acquireKlineWriterLockSync({ ...options, owner });
  try {
    return callback();
  } finally {
    releaseKlineWriterLockSync(lock);
  }
}

File: scripts/kline_sqlite_durability.py
#!/usr/bin/env python3
"""NFS-safe SQLite durability and cross-language writer coordination for Kline data."""

from __future__ import annotations

from contextlib import contextmanager
import json
import os
from pathlib import Path
import tempfile
import time
import uuid


DEFAULT_LOCK_FILE = Path(
    os.environ.get(
        "KLINE_SQLITE_WRITER_LOCK_FILE",
        str(Path(tempfile.gettempdir()) / "kline_sqlite_single_writer.lock"),
    )
)
DEFAULT_TIMEOUT_SEC = float(os.environ.get("KLINE_SQLITE_WRITER_LOCK_TIMEOUT_SEC", "30"))
POLL_SEC = float(os.environ.get("KLINE_SQLITE_WRITER_LOCK_POLL_SEC", "0.025"))
INVALID_LOCK_GRACE_SEC = 5.0


def _process_is_alive(pid: object) -> bool:
    try:
        parsed = int(pid)
    except (TypeError, ValueError):
        return False
    if parsed <= 0:
        return False
    try:
        os.kill(parsed, 0)
        return True
    except PermissionError:
        return True
    except ProcessLookupError:
        return False


def _read_owner(lock_file: Path) -> dict | None:
    try:
        value = json.loads(lock_file.read_text(encoding="utf-8"))
        return value if isinstance(value, dict) else None
    except (OSError, json.JSONDecodeError):
        return None


def _stale_reason(lock_file: Path, owner: dict | None) -> str | None:
    if owner and owner.get("pid") and not _process_is_alive(owner.get("pid")):
        return "owner_process_not_alive"
    if owner:
        return None
    try:
        if time.time() - lock_file.stat().st_mtime >= INVALID_LOCK_GRACE_SEC:
            return "invalid_owner_record"
    except FileNotFoundError:
        return None
    return None


def _same_owner(expected: dict | None, current: dict | None) -> bool:
    if (expected or {}).get("token") or (current or {}).get("token"):
        return (expected or {}).get("token") == (current or {}).get("token")
    if (expected or {}).get("pid") or (current or {}).get("pid"):
        return (
            (expected or {}).get("pid") == (current or {}).get("pid")
            and (expected or {}).get("owner") == (current or {}).get("owner")
        )
    return expected is None and current is None


def _try_cleanup_gate(lock_file: Path) -> tuple[Path, int, str] | None:
    cleanup = Path(f"{lock_file}.cleanup")
    token = str(uuid.uuid4())
    fd = None
    try:
        fd = os.open(cleanup, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        payload = json.dumps(
            {
                "schema_version": "kline_sqlite_cleanup_gate.v1",
                "token": token,
                "pid": os.getpid(),
                "acquired_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            },
            sort_keys=True,
        ).encode("utf-8")
        os.write(fd, payload)
        os.fsync(fd)
        return cleanup, fd, token
    except FileExistsError:
        return None
    except BaseException:
        if fd is not None:
            os.close(fd)
            try:
                cleanup.unlink()
            except FileNotFoundError:
                pass
        raise


def _release_cleanup_gate(gate: tuple[Path, int, str] | None) -> None:
    if not gate:
        return
    cleanup, fd, token = gate
    try:
        owner = _read_owner(cleanup)
        if owner and owner.get("token") == token:
            try:
                cleanup.unlink()
            except FileNotFoundError:
                pass
    finally:
        os.close(fd)


def _remove_stale_lock(lock_file: Path, observed_owner: dict | None) -> bool:
    gate = _try_cleanup_gate(lock_file)
    if not gate:
        return False
    try:
        current_owner = _read_owner(lock_file)
        reason = _stale_reason(lock_file, current_owner)
        if not reason or not _same_owner(observed_owner, current_owner):
            return False

        stale = lock_file.with_name(
            f"{lock_file.name}.stale-{int(time.time() * 1000)}-{os.getpid()}-{uuid.uuid4()}"
        )
        try:
            lock_file.replace(stale)
        except FileNotFoundError:
            return False
        try:
            stale.unlink()
        except FileNotFoundError:
            pass
        print(
            f"[kline-sqlite] removed stale writer lock ({reason}): {lock_file}",
            file=os.sys.stderr,
        )
        return True
    finally:
        _release_cleanup_gate(gate)


@contextmanager
def kline_single_writer(
    name: str,
    *,
    lock_file: str | os.PathLike | None = None,
    timeout_sec: float | None = None,
):
    """Coordinate with Node Kline writers through one atomic local lock file."""

    path = Path(
        lock_file
        or os.environ.get("KLINE_SQLITE_WRITER_LOCK_FILE")
        or DEFAULT_LOCK_FILE
    )
    path.parent.mkdir(parents=True, exist_ok=True)
    timeout = DEFAULT_TIMEOUT_SEC if timeout_sec is None else max(0.001, float(timeout_sec))
    deadline = time.monotonic() + timeout
    token = str(uuid.uuid4())
    fd = None

    while True:
        try:
            fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
            payload = json.dumps(
                {
                    "schema_version": "kline_sqlite_writer_lock.v1",
                    "token": token,
                    "pid": os.getpid(),
                    "owner": str(name or "unknown"),
                    "acquired_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                },
                sort_keys=True,
            ).encode("utf-8")
            os.write(fd, payload)
            os.fsync(fd)
            break
        except FileExistsError:
            owner = _read_owner(path)
            stale_reason = _stale_reason(path, owner)
            if stale_reason and _remove_stale_lock(path, owner):
                continue
            if time.monotonic() >= deadline:
                holder = (
                    f"pid={owner.get('pid', 'unknown')} owner={owner.get('owner', 'unknown')}"
                    if owner
                    else "owner=initializing"
                )
                raise TimeoutError(
                    f"timed out waiting for Kline SQLite writer lock {path} ({holder})"
                )
            time.sleep(min(POLL_SEC, max(0.001, deadline - time.monotonic())))

    try:
        yield
    finally:
        try:
            owner = _read_owner(path)
            if owner and owner.get("token") == token:
                try:
                    path.unlink()
                except FileNotFoundError:
                    pass
        finally:
            if fd is not None:
                os.close(fd)


def configure_kline_sqlite_connection(db, *, busy_timeout_ms: int | None = None) -> dict:
    """Apply and verify the Kline rollback-journal contract on one writer connection."""

    timeout_ms = int(
        busy_timeout_ms
        if busy_timeout_ms is not None
        else os.environ.get("KLINE_SQLITE_BUSY_TIMEOUT_MS", "30000")
    )
    db.execute(f"PRAGMA busy_timeout = {max(1, timeout_ms)}")
    db.execute("PRAGMA mmap_size = 0")
    row = db.execute("PRAGMA journal_mode = DELETE").fetchone()
    journal_mode = str(row[0] if row else "").strip().upper()
    db.execute("PRAGMA synchronous = FULL")
    synchronous = int(db.execute("PRAGMA synchronous").fetchone()[0])
    mmap_size = int(db.execute("PRAGMA mmap_size").fetchone()[0])
    if journal_mode != "DELETE" or synchronous != 2 or mmap_size != 0:
        raise RuntimeError(
            "unsafe Kline SQLite settings: "
            f"journal_mode={journal_mode or 'unknown'} "
            f"synchronous={synchronous} mmap_size={mmap_size}"
        )
    return {
        "journal_mode": journal_mode,
        "synchronous": "FULL",
        "mmap_size": mmap_size,
        "busy_timeout_ms": max(1, timeout_ms),
    }
